gcloud/pubsub/client.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2015 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from gcloud._helpers import get_default_project
+from gcloud._helpers import _ClientProxy
+from gcloud.pubsub._implicit_environ import _require_connection
+from gcloud.pubsub import api
+from gcloud.pubsub.subscription import Subscription
+from gcloud.pubsub.topic import Topic
+
+
+class Client(object):
+    """Wrap :mod:`gcloud.pubsub` API objects.
+
+    :type connection: :class:`gcloud.pubsub.connection.Connection` or None
+    :param connection: The configured connection. Defaults to one inferred
+                       from the environment.
+
+    :type project: str or None
+    :param connection: The configured project. Defaults to the value inferred
+                       from the environment.
+    """
+
+    def __init__(self, connection=None, project=None):
+        self.connection = _require_connection(connection)
+        if project is None:
+            project = get_default_project()
+        self.project = project
+
+    def topic(self, name):
+        """Proxy for :class:`gcloud.pubsub.topic.Topic`.
+
+        :type name: string
+        :param name: the name of the topic
+
+        :rtype: :class:`_Topic`
+        :returns: a proxy for a newly created Topic, using the passed name
+                  and the client's project.
+        """
+        topic = Topic(name, self.project)
+        return _Topic(topic, self)
+
+    def list_topics(self, page_size=None, page_token=None):
+        """Proxy for :func:`gcloud.pubsub.api.list_topics`.
+
+        Passes configured connection and project.
+        """
+        topics, next_page_token = api.list_topics(
+            page_size=page_size,
+            page_token=page_token,
+            connection=self.connection,
+            project=self.project)
+        proxies = [_Topic(topic, self) for topic in topics]
+        return proxies, next_page_token
```

Comment:
So the usage of getting all the topics here is:

```python
topics, next_page_token = client.list_topics()
more_topics = client.list_topics(page_token=next_page_token)
# ...
```

or 

```python
all_topics = []

topics, token = client.list_topics()
all_topics.extend(topics)
while token is not None:
  topics, token = client.list_topics(page_token=token)
  all_topics.extend(topics)
```

?

Could we return an iterator so that the usage looks like...

```python
all_topics = list(client.list_topics())
```

or so that I could do...

```python
for topic in client.list_topics():
  if topic.name == 'my-topic':
    break  # Only navigate through as many pages as needed...
```